File: hw2/nqueens.py
import time
from random import *
import queue
from collections import Counter

class n_queens:
    def __init__(self, N, queen_pos):
        self.N = N  # SIZE OF BOARD
        self.queen_pos = queen_pos #ARR[i] RETURNS ROW POS OF COL I QUEEN
        self.nodes_visited = 0

        # fill queen pos array
        for i in range(0,N):
            queen_pos.append(-2*N) #fill array with junk for now

    
    def checkPositions(self, row, col):
        # given a queen in column 'col' and in row 'row', is it valid vs the rest of the board
        row_to_check = row
        diag1_check = row - col
        diag2_check = row + col

        # loop thru queen pos array to validate positions
        for c in range(0, len(self.queen_pos) ):
            if c == col:
                continue #dont want to validate against self
            cur_row = self.queen_pos[c]
            cur_diag1 = self.queen_pos[c] - c
            cur_diag2 = self.queen_pos[c] + c

            if row_to_check == cur_row or diag1_check == cur_diag1 or diag2_check == cur_diag2:
                return False

        return True

    def print_sol(self):
        res = [ [0] * self.N for _ in range(self.N) ]
        for i in range(0, self.N):
            res[i][self.queen_pos[i]] = self.queen_pos[i]+1
        for j in range(0, self.N):
            s = ""
            for k in range(0, self.N):
                if res[j][k] == 0:
                    s += "- " #print("- ")
                else:
                    s+= str(res[j][k]) + " " #print(str(res[j][k]) + " ")
            print(s)
        return

class backtracking(n_queens):
    def __init__(self, N, queen_pos):
        n_queens.__init__(self, N, queen_pos)
    
    def backtrack(self, col): # col is current col being checked
        self.nodes_visited += 1
        if self.N==2 or self.N==3:
            print("NO SOLUTIONS EXIST FOR THE GIVEN N")
            return False #no sols. exist for n=2 or 3
        if self.N==col:
            self.print_sol()
            return True #done
        
        for r in range(0, self.N,): #loop over every row trying to get a valid solution for column 'col'
            self.queen_pos[col] = r
            if self.checkPositions(r, col) == True:
                if self.backtrack(col+1) == True:
                    return True
            else:
                self.queen_pos[col] = -2*self.N; #failed, go back to def value


class forwardtracking(n_queens):
    def __init__(self, N, queen_pos):
        n_queens.__init__(self, N, queen_pos)
        self.domains = dict()
        for i in range(N):
            temp = domain_values(N,i)
            self.domains[i]=temp.domain

    def domain_wipeout(self, old_domains, row, col):
        #1. eliminate rows
        #2. eliminate r-c, diag1
        #3. eliminate r+c, diag2

        # old_domains is a copy of current stacks current domains

        #1. rows
        for i in range(self.N):
            if i==col:
                continue
            old_domains[i] = (old_domains[i]).difference({row})

        #2. diag1, row-col
        dif = row - col
        for i in range(self.N):
            if i==col:
                continue
            old_domains[i] = (old_domains[i]).difference({i+dif})
        
        #3. diag2, row+col
        add = row + col
        for i in range(self.N):
            if i==col:
                continue
            old_domains[i] = (old_domains[i]).difference({add-i})

        return old_domains


    def forwardcheck(self,col, domains): #pretty much same as back track, domains will be copy of previous stack's domains
        self.nodes_visited += 1
        # invalid game size 2 or 3.
        if  self.N == 2 or self.N == 3:
            return False
        # if N == col, N queens finished
        if self.N == col:
            self.print_sol()
            return True
        # if curent columns domain size is zero, backtrack
        if len(domains[col])==0:
            return False
        
        # while there is still a row to try in current domain
        while (domains[col]).__len__() >0 :
            r = (domains[col]).pop()
            # assume current row is good
            self.queen_pos[col] = r
            # if valid
            if self.checkPositions(r, col) == True:
                # domain wipeout!
                temp = domains.copy()
                new_domains = self.domain_wipeout(temp, r, col )
                # if any of the domains are 0, kill stack
                valid = True
                for i in range(col+1,self.N):
                    if new_domains[i].__len__() == 0 :
                        valid = False
                if not valid:
                    break
                # print("CUR COL : " + str(col))
                if self.forwardcheck(col+1, new_domains) == True:
                    return True
            else: 
                self.queen_pos[col] = -2*self.N

class HillClimb:
    def __init__(self, N):
        self.N = N #nxN board
        
    def initial_state(self): #randomly creates board setup
        return list( randrange(self.N) for i in range(self.N) )

    def goal_state(self, current_state): #is current state a valid one?
        c1, c2, c3 =  ( set() for i in range(3) ) #checks for col, diag1 and diag2
        for row, col in enumerate(current_state):
            if col in c1 or row - col in c2 or row + col in c3:
                return False
            c1.add(col)
            c2.add(row - col)
            c3.add(row + col)
        return True

    def h(self, current_state): #h(n), heuristic for # of queens being attacked
        c1, c2, c3 = [Counter() for i in range(3)] # counts for row, diag1, and diag2 being attacked
        for row,col in enumerate(current_state):
            c1[row] = c1[row] + 1
            c2[row - col] = c2[row - col] + 1
            c3[row + col] = c3[row + col] + 1

        collisions = 0
        for c in [c1, c2, c3]:
            for k in c:
                collisions += c[k] * (c[k]-1)/2 #Counter[x] gives number of counts of key X. This will give us total collisions, not incuding colision vs self
        return -collisions 
    
    def children(self, current_state):
        children = []
        for r in range(self.N):
            for c in range(self.N):
                if c != current_state[r]: #if no queen exists for a certain col, append to a list of queens
                    temp = list(current_state)
                    temp[r] = c
                    children.append(list(temp))

        return children

def h_c(state):
    current = state.initial_state()
    while True:
        children = state.children(current)
        if not children: #no where else to go !
            break
        # get worst child
        child = max(children, key=lambda st: state.h(st))
        if state.h(child) <= state.h(current):
            break    
        current = child
    return current

def main():
    print("n_queens mulvey\n")
    n = int(input("ENTER size of board ? : "))
    # print("\n====!!BACKTRACKING!!====")
    # start_time = time.time()
    # a = backtracking(n, [])
    # a.backtrack(0)
    # end_time = time.time() - start_time
    # print("BACKTRACKING took " + str(end_time) + "secs to run (and print) with " + str(a.nodes_visited) + " nodes visited")
    # print("\n\n====!!FORWARD-CHECING!!====")
    # start_time = time.time()
    # b = forwardtracking(n, [])
    # b.forwardcheck(0, b.domains)
    # end_time = time.time() - start_time
    # print("FORWARDCHECKING took " + str(end_time) + "secs to run (and print) with " + str(b.nodes_visited) + " nodes visited")
    d = HillClimb(n)
    ans = h_c(d)
    print(ans)
    return

#-----#
main()

File: hw2/N_Queens.java
// author : tom mulvey

import java.util.*;

public class N_Queens {
    /*
        @@@ given a number, N, place N queens on a NXN chess board such that they wont attack eachother.
            a) Implement the backtracking algorithm to solve the n-queens problem.
            b) Add forward checking to the backtracking algorithm. Compare backtracking alone versus backtracking with forward checking for an increasing number of queens. 
            Turn in: Cover page, pseudo-code of your algorithm, code, sample outputs, comparison results, conclusion. 
    
        variables: Queen_0 -> Queen_N-1
        domain: rows 0->N-1
        constraints: chess rules. 
            (queen 0 is in the first col, q1 in second, etc)
    */

    public int N; // board size and number of queens
    public int[] queen_positions; //index 0 contains queen0(col)'s row.

    // DEFAULT CONSTRUCTOR
    public N_Queens(){
        // init size N
        Scanner sc = new Scanner(System.in);
        System.out.print("Please enter N: ");
        this.N = sc.nextInt();
        //close sc
        sc.close();
        String x = "---" + this.N + " Queens!---";
        System.out.println(x);

        // explicitly set array to all zeros
        this.queen_positions = new int[this.N];
        Arrays.fill(this.queen_positions, -2*this.N);//if we fill arrays with 0 then it will fail on checking col zero pos zero

        // FOR FORWARD CHECKING WE NEED DOMAINS
        // we will have an map of sets. q_domain[0] will give us the domain of first queen.
        // HashMap<Integer, HashSet<Integer>> q_domains = new HashMap<Integer, HashSet<Integer>>();
        // // fill hashmap with copies of domains
        // HashSet<Integer> set1 = new HashSet<Integer>();
        // for( int i=0; i<this.N; i++){
        //     set1.add(i);
        // }
        // HashSet<Integer> Clone = new HashSet<Integer>()
        // q_domains.put( 0, (HashSet)set1.clone() );
        // // Set<Integer> set2 = set1.stream().map(Integer::new).collect(Collectors.toSet());
    }

    // constructor with intended size
    public N_Queens(int size){
        this.N = size;
        // explicitly fill arr with all zeros
        this.queen_positions = new int[size];
        Arrays.fill(this.queen_positions, -2*this.N);//if we fill arrays with 0 then it will fail on checking col zero pos zero
    }

    public void print_sol() {
        int[][] res = new int[this.N][this.N];
        for (int i = 0; i < this.N; i++) {
            res[i][this.queen_positions[i]] = this.queen_positions[i]+1;
        }
        for (int i = 0; i < this.N; i++) {
            for (int j = 0; j < this.N; j++) {
                if (res[i][j] == 0){
                    System.out.print("-" + " ");
                } else {
                    System.out.print(res[i][j] + " ");
                }
            }
            System.out.println();
        }
        System.out.println("\n");
    }

    public boolean checkPositions(int col, int r){ //=O(N)
        // int col is current column, r is the row we are trying to place on
        // since each index/col of queen arrary is different, we only need to check rows and diagonals
        /* diag 1:
             0 1 2 3
           0   -
           1     Q
           2       -
           3  
             Queen row-col = -1. Thus any other queens row - col cannot = -1. The other diag is adding row+col.
             Also need to make sure no other queen as the same row value. For all i,j in arr, i!=j
           */

        int row_check = r;
        int diag1_check = r - col;
        int diag2_check = col + r;

        for (int i=0; i<this.N; i++) { // loop thru queens pos array to find errors shown above
            // we are given current column, check queens_position[col] vs all other entries
            if(i == col)
                continue;
            
            int i_row = this.queen_positions[i]; 
            int i_diag1 = this.queen_positions[i] - i;
            int i_diag2 = i + this.queen_positions[i];

            if( row_check == i_row || diag1_check == i_diag1 || diag2_check == i_diag2 ){
                return false;
            }
        }

        // we made it! very nice! 
        return true;
    }

    public boolean backtracking(int n, int col){
    // col is current column being tested in backtracking.
    // IMPLEMENTATION OF BACKTRACKING ALGORITHM
        if( n==2 || n==3 ){
            return false; // no solution exists for cases n=2 or n=3
        }
        if( col == n){
            return true;
        }

        for( int r=0; r<n; r++ ){ //loop over every row trying to get a valid solution for column col
            this.queen_positions[col] = r;
            //System.out.println("Checking queen" + col + " at row " + r);
            if( checkPositions(col, r) ){
                //System.out.println("!!!!queen" + col + " at pos " + r + " TRUEE!!!!");
                if( backtracking(n, col+1 ) )
                    return true;       
            } else {
                this.queen_positions[col] = -2*this.N; //failed, go back to def value
            }
        }

        return false;
    }


    public boolean forward_checking(int n, int col, int[][] domains){
        if( n==2 || n==3 ){
            return false; // no solution exists for cases n=2 or n=3
        }
        if( col == n){
            return true;
        }

        // 1. check ALL domain sizes.
        // 2. find value
        // 3. Copy domains, domain wipeout the copies
        // 4. go to next queen


        return false;
    }


    public static void main(String[] args) {
        N_Queens game = new N_Queens();
        System.out.println(game.N + "Queens solution: " + game.backtracking(game.N, 0));
        game.print_sol();
        
        System.out.println(Arrays.toString(game.queen_positions));
    }
}